tests/local_db/test_data_access.py

```diff
@@ -128,6 +128,34 @@ def test_withdraw_file_from_request_bad_state(status):
         )
 
 
+def test_withdraw_file_from_request_file_does_not_exist():
+    author = factories.create_user(username="author", workspaces=["workspace"])
+    release_request = factories.create_request_at_status(
+        "workspace",
+        author=author,
+        status=RequestStatus.RETURNED,
+        files=[factories.request_file(changes_requested=True, path="foo.txt")],
+    )
+
+    # foo.txt does not exist and can't be withdrawn
+    with pytest.raises(exceptions.FileNotFound):
+        dal.withdraw_file_from_request(
+            release_request.id,
+            UrlPath("bar.txt"),
+            AuditEvent.from_request(
+                release_request, AuditEventType.REQUEST_FILE_WITHDRAW, user=author
+            ),
+        )
+    # foo.txt can be withdrawn
+    dal.withdraw_file_from_request(
+        release_request.id,
+        UrlPath("foo.txt"),
+        AuditEvent.from_request(
+            release_request, AuditEventType.REQUEST_FILE_WITHDRAW, user=author
+        ),
+    )
+
```

Comment:
would be nice to see a test case here for a file that has already been withdrawn, but as we have coverage it's presumably covered elsewhere?

Reply:
It was implicitly tested elsewhere, but I've added a specific test for it now (in test_business_logic.py). The permission is enforced in business_logic.py, so we shouldn't ever get to attempting to withdraw and already withdrawn file from the dal (and it wouldn't care anyway, it would just be trying to set the file status again)